pkg/operator/resource/resourceapply/apps.go

```diff
@@ -15,13 +18,42 @@ import (
 	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
 )
 
-// ApplyDeployment merges objectmeta and requires matching generation. It returns the final Object, whether any change as made, and an error
-func ApplyDeployment(client appsclientv1.DeploymentsGetter, recorder events.Recorder, required *appsv1.Deployment, expectedGeneration int64,
+const specHashAnnotation = "operator.openshift.io/spec-hash"
+
+// ApplyDeployment merges ObjectMeta of the provided deployment with an existing one
+// if it exists and updates the API if the deployment spec and metadata differ. To
+// ensure an update in response to state external to the deployment spec, the caller
+// should set an annotation representing that external state.
+func ApplyDeployment(client appsclientv1.DeploymentsGetter, recorder events.Recorder,
+	required *appsv1.Deployment) (*appsv1.Deployment, bool, error) {
+
+	// Record the hash of the spec in an annotation to prompt rollout if it changes.
+	jsonBytes, err := json.Marshal(required.Spec)
+	if err != nil {
+		return nil, false, err
+	}
+	specHash := fmt.Sprintf("%x", sha256.Sum256(jsonBytes))
```

Comment:
The deployment controller creates the replicaSet with a hash in its name. This hash is calculated from the deployment's pod template [0]. Same for the daemonSet controller.

It's not the same as we're doing here, as we're hashing the whole deployment's spec, but it'd be nice to use a similar approach. For that, we could either:

1) Hash the `deployment.Spec.Template` as well, and use the public available utility function [1] that does that. The downside is that `ApplyDeployment()` wouldn't update the deployment if we change the number of replicas, for instance.

2) Create our own hashing function based on [1].

Does this make sense?

[0] https://github.com/kubernetes/kubernetes/blob/98e65951dccfd40d3b4f31949c2ab8df5912d93e/pkg/controller/deployment/sync.go#L189
[1] https://github.com/kubernetes/kubernetes/blob/98e65951dccfd40d3b4f31949c2ab8df5912d93e/pkg/controller/controller_utils.go#L1129

Reply:
deployment controller: hash to come up with a unique replicaset name
operator: hash to ensure that a change in intent results in a rollout

Given that the use cases are different, why would it be desirable to match the implementation of the deployment controller?